src/e84_geoai_common/llm/batch.py

```diff
@@ -0,0 +1,324 @@
+from e84_geoai_common.llm.models.claude import (
+    CLAUDE_BEDROCK_MODEL_IDS,
+    BedrockClaudeLLM,
+)
+from e84_geoai_common.llm.models.nova import BedrockNovaLLM
+
+if TYPE_CHECKING:
+    from mypy_boto3_s3.literals import BucketLocationConstraintType
+
+# Batch inference uses camel case for its variables. Ignore any linting problems with this.
+# ruff: noqa: N815
+
+ValidBatchLLMs = BedrockClaudeLLM | BedrockNovaLLM
+
+
+class BatchRecordInput[RequestModel: BaseModel](BaseModel):
+    """Single input record for batch inference."""
+
+    recordId: str
+    modelInput: RequestModel
+
+
+class BatchRecordOutput[RequestModel: BaseModel, ResponseModel: BaseModel](BaseModel):
+    """Specific output record for Claude batch inference."""
+
+    model_config = ConfigDict(strict=True, extra="forbid", frozen=True)
+
+    recordId: str
+    modelInput: RequestModel
+    modelOutput: ResponseModel | None = None
+    error: RequestModel | None = None
+
+
+class S3InputDataConfig(BaseModel):
+    """S3 input URI model."""
+
+    s3Uri: str
+
+
+class S3OutputDataConfig(BaseModel):
+    """S3 output URI model."""
+
+    s3Uri: str
+
+
+class InputDataConfig(BaseModel):
+    """Wrapper for S3 input URI model."""
+
+    s3InputDataConfig: S3InputDataConfig
+
+
+class OutputDataConfig(BaseModel):
+    """Wrapper for S3 output URI modoel."""
+
+    s3OutputDataConfig: S3OutputDataConfig
+
+
+class BatchLLMRequest(BaseModel):
+    """Request payload for invoking a batch model job."""
+
+    roleArn: str
+    jobName: str
+    modelId: str
+    inputDataConfig: InputDataConfig
+    outputDataConfig: OutputDataConfig
+
+
+class BedrockBatchInference[RequestModel: BaseModel, ResponseModel: BaseModel]:
+    def __init__(
+        self,
+        llm: ValidBatchLLMs,
+        request_model: type[RequestModel],
+        response_model: type[ResponseModel],
+        bedrock_client: BedrockClient | None = None,
+        s3_client: S3Client | None = None,
+    ) -> None:
+        """Initalizes BedrockBatchInference.
+
+        Args:
+            llm: valid LLM instance.
+            request_model: Valid InvokeLLMRequest model (i.e. ClaudeInvokeLLMRequest)
+            response_model: Valid Response model (i.e. ClaudeResponse)
+            bedrock_client: Optional pre-initialized bedrock boto3 client. Defaults to None.
+            s3_client: Optional pre-initialized s3 boto3 client. Defaults to None.
+        """
+        self.llm = llm or BedrockClaudeLLM(model_id=CLAUDE_BEDROCK_MODEL_IDS["Claude 3.5 Haiku"])
+        self.request_model = request_model
+        self.response_model = response_model
+        self.bedrock_client = bedrock_client or boto3.client("bedrock")  # type: ignore[reportUnknownMemberType]
+        self.s3_client = s3_client or boto3.client("s3")  # type: ignore[reportUnknownMemberType]
+        self.batch_record_output_model = BatchRecordOutput[self.request_model, self.response_model]
+
+    def get_results(
+        self: Self, job_arn: str
```

Comment:
While not technically incorrect, it is not common practice to annotate the `self` param in methods with the `Self` type. That type is mostly for use as the return type of methods such that it works well with inheritance. E.g.

```py
class Shape:
    def rescale(self, factor: float) -> Self:
        ...

class Circle(Shape):
    ...
```

Here, thanks to the magic of `Self`, the return type of `Circle.rescale()` will be inferred to be `Circle` without having to be explicitly declared as such.